topk-js/src/query/query.rs

```diff
@@ -0,0 +1,157 @@
+use std::collections::HashMap;
+
+use napi::bindgen_prelude::*;
+use napi_derive::napi;
+
+use crate::{
+    data::scalar::Scalar,
+    expr::{
+        filter::FilterExpressionUnion,
+        logical::{LogicalExpression, LogicalExpressionUnion},
+        select::SelectExpression,
+    },
+};
+
+use super::stage::Stage;
+
+#[napi(namespace = "query")]
+#[derive(Debug, Clone)]
+pub struct Query {
+    stages: Vec<Stage>,
+}
+
+#[napi(namespace = "query")]
+impl Query {
+    #[napi(constructor)]
+    pub fn new() -> Self {
+        Self { stages: vec![] }
+    }
+
+    #[napi(factory)]
+    pub fn select(
+        &self,
+        #[napi(ts_arg_type = "Record<string, LogicalExpression | FunctionExpression>")]
+        exprs: HashMap<String, SelectExpression>,
+    ) -> Self {
+        let mut new_query = Query {
+            stages: self.stages.clone(),
+        };
+
+        let stage = Stage::Select {
+            exprs: exprs.into_iter().map(|(k, v)| (k, v.into())).collect(),
+        };
+
+        new_query.stages.push(stage);
+
+        new_query
+    }
+
+    #[napi]
+    pub fn filter(
+        &self,
+        #[napi(ts_arg_type = "LogicalExpression | TextExpression")] expr: FilterExpressionUnion,
+    ) -> Query {
+        let mut new_query = Query {
+            stages: self.stages.clone(),
+        };
+
+        new_query.stages.push(Stage::Filter { expr });
+
+        new_query
+    }
+
+    #[napi]
+    pub fn top_k(&self, expr: LogicalExpression, k: i32, asc: Option<bool>) -> Query {
+        let mut new_query = Query {
+            stages: self.stages.clone(),
+        };
+
+        new_query.stages.push(Stage::TopK {
+            expr,
+            k,
+            asc: asc.unwrap_or(false),
+        });
+
+        new_query
+    }
+
+    #[napi]
+    pub fn count(&self) -> Query {
+        let mut new_query = Query {
+            stages: self.stages.clone(),
+        };
+
+        new_query.stages.push(Stage::Count {});
+
+        new_query
+    }
+
+    #[napi]
+    pub fn rerank(
+        &self,
+        model: Option<String>,
+        query: Option<String>,
+        fields: Option<Vec<String>>,
+        topk_multiple: Option<u32>,
+    ) -> Query {
+        let mut new_query = Query {
+            stages: self.stages.clone(),
+        };
+
+        new_query.stages.push(Stage::Rerank {
+            model,
+            query,
+            fields: fields.unwrap_or_default(),
+            topk_multiple,
+        });
+
+        new_query
+    }
+}
+
+#[napi(namespace = "query")]
+pub fn select(
```

Comment:
we should also export `filter()`, `top_k` and `count` query "starters", like we in python. `rerank()` on it's own doesn't make sense.